### PR TITLE
Warn about exit in helper

### PR DESCRIPTION
**Pull request by jaredbeck**
In https://github.com/rspec/rspec-core/issues/2508, @JonRowe wondered:

> Yes using `exit` or `abort` will cause RSpec to quit .. I am wondering if we should print a warning about that happening...

I said:

> Yes, please! Such a warning will be very helpful, as in https://github.com/rspec/rspec-rails/issues/2441#issuecomment-1003212668 The current output "No examples found" is misleading.

Then, @pirj asked:

> Does anyone want to send a PR?

So, this is what you get. 😅  I'm sure it needs a lot of work. 🤦  If it's at all salvageable, please help me get it cleaned up.

### CODE REVIEW

File: lib/rspec/core/configuration.rb
Comment (JonRowe) [inline lib/rspec/core/configuration.rb:2137]: How about?
```suggestion
        reporter.notify_non_example_exception(ex, "While loading #{relative_file} an `exit` / `raise SystemExit` occurred, RSpec will now quit.")
```

File: spec/rspec/core/configuration_spec.rb
Comment (JonRowe) [inline spec/rspec/core/configuration_spec.rb:273]: This isn't needed really, you have an integration test which is more useful.

File: spec/integration/spec_file_load_errors_spec.rb
Comment (JonRowe) [inline spec/integration/spec_file_load_errors_spec.rb:94]: I think this need updating to match :)

File: spec/integration/spec_file_load_errors_spec.rb
Comment (JonRowe) [inline spec/integration/spec_file_load_errors_spec.rb:101]: I wonder if we can supress this, maybe setting a global `rspec_is_quitting` in config and gateing this message around that?

File: lib/rspec/core/world.rb
Comment (JonRowe) [inline lib/rspec/core/world.rb:197]: The message we want to surpress is the false "no examples found", theres already an explanation of whats happening above this, and actually if a SystemExit is encountered during a spec run there would be an example count.

```suggestion
        if filter_manager.empty? && !rspec_is_quitting
```

Comment (jaredbeck) [reply lib/rspec/core/world.rb:197]: > Sugesstion: `if filter_manager.empty? && !rspec_is_quitting`

I considered that, but In my testing, it produced the output: "All examples were filtered out", which seemed misleading.

> .. if a SystemExit is encountered during a spec run there would be an example count.

In the scenario that I'm testing, the `exit` occurs in a `--require` (e.g. at the top of 'spec_helper'). Would there still be a count in that case?

Comment (JonRowe) [reply lib/rspec/core/world.rb:197]: Hm fair, prehaps it should be:
```suggestion     
        unless rspec_is_quitting
          if filter_manager.empty?
            report_filter_message("No examples found.")
          elsif exclusion_filter.empty? || inclusion_filter.empty?
            report_filter_message(everything_filtered_message)
          end
        end
```

**Review by JonRowe — APPROVED** [review]

**Comment by JonRowe** [comment]
Thanks for this, just need to wait for the build to pass and then I'll merge it.

File: spec/integration/spec_file_load_errors_spec.rb
Comment (JonRowe) [inline spec/integration/spec_file_load_errors_spec.rb:100]: This output needs to be adapted on JRuby as it comes out differently (from the build):
```diff
Diff:

       @@ -1,9 +1,8 @@
        
        While loading ./helper_with_exit.rb an `exit` / `raise SystemExit` occurred, RSpec will now quit.
       -Failure/Error: exit 999
       +Failure/Error: Unable to find org/jruby/RubyKernel.java to read failed line
        
        SystemExit:
          exit
       -# ./helper_with_exit.rb:1:in `exit'
        # ./helper_with_exit.rb:1:in `<main>'
```
Can you adjust this spec to cope with that?

Comment (JonRowe) [reply spec/integration/spec_file_load_errors_spec.rb:107]: This diff is still slightly wrong, line 106 `# ./helper_with_exit.rb:1:in `exit'` is missing entirely on Jruby too

**Comment by jaredbeck** [comment]
> This diff is still slightly wrong, line 106 # ./helper_with_exit.rb:1:in exit'` is missing entirely on Jruby too

Sorry for the back-and-forth. I can't (don't want to) run jruby locally, and GH says "workflow awaiting approval" so I'm just taking shots in the dark. In case`c85bacc` doesn't work, is there some way you can tell GH to run CI for me?

**Comment by JonRowe** [comment]
> Sorry for the back-and-forth. I can't (don't want to) run jruby locally, and GH says "workflow awaiting approval" so I'm just taking shots in the dark. In case c85bacc doesn't work, is there some way you can tell GH to run CI for me?

It's no worries, I've been checking on it when I see an update and approving, Github doesn't remember previous approvals on the first PR sadly, future PRs should remember we approved you.

**Comment by JonRowe** [comment]
Merging despite build failure as it is segfaults for windows and ruby-head.

**Pull request merged**
